File: crates/aily-core/src/ally.rs
//! Local propose-only day planning (deterministic, no model).
//! Mirrors the web `ally.js` contract for shared honesty.

const FLOOR: u32 = 15;

#[derive(Debug, Clone)]
pub struct TargetHint {
    pub id: String,
    pub title: String,
    /// 0.0 ..= 1.0 journey progress (higher = closer to done).
    pub progress: f64,
    pub soft_hours: Option<f64>,
}

#[derive(Debug, Clone)]
pub struct ExistingCommitment {
    pub id: String,
    pub target_id: String,
    pub estimate_min: u32,
    pub must_keep: bool,
    pub status: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Proposal {
    pub text: String,
    pub target_id: String,
    pub estimate_min: u32,
    pub must_keep: bool,
    pub reason: String,
}

#[derive(Debug, Clone)]
pub struct ProposeInput {
    pub weekly_capacity_hours: f64,
    pub nights_per_week: f64,
    pub targets: Vec<TargetHint>,
    pub soft_caps: Vec<(String, f64)>,
    pub existing_today: Vec<ExistingCommitment>,
    pub intention: String,
    pub max_items: usize,
}

#[derive(Debug, Clone)]
pub struct ProposeOutput {
    pub ok: bool,
    pub proposals: Vec<Proposal>,
    pub summary: String,
    pub remaining_min: u32,
    pub error: Option<String>,
}

fn clamp_estimate(min: f64) -> u32 {
    if !min.is_finite() || min < FLOOR as f64 {
        return FLOOR;
    }
    let snapped = ((min / 15.0).round() as u32) * 15;
    snapped.max(FLOOR)
}

fn daily_soft_cap_minutes(weekly_hours: f64, nights: f64) -> f64 {
    let n = nights.max(1.0);
    (weekly_hours * 60.0 / n).max(15.0)
}

fn day_fits(
    weekly_hours: f64,
    nights: f64,
    soft_caps: &[(String, f64)],
    existing: &[ExistingCommitment],
    proposals: &[Proposal],
    next: &Proposal,
) -> bool {
    let daily_cap = daily_soft_cap_minutes(weekly_hours, nights);
    let mut total = 0.0_f64;
    let mut per_goal: std::collections::HashMap<String, f64> = std::collections::HashMap::new();

    let push = |tid: &str, mins: f64, total: &mut f64, per: &mut std::collections::HashMap<String, f64>| {
        *total += mins;
        *per.entry(tid.to_string()).or_insert(0.0) += mins;
    };

    for c in existing.iter().filter(|c| c.status != "dropped") {
        push(
            &c.target_id,
            c.estimate_min as f64,
            &mut total,
            &mut per_goal,
        );
    }
    for p in proposals {
        push(&p.target_id, p.estimate_min as f64, &mut total, &mut per_goal);
    }
    push(
        &next.target_id,
        next.estimate_min as f64,
        &mut total,
        &mut per_goal,
    );

    if total > daily_cap + 1e-9 {
        return false;
    }
    if soft_caps.len() >= 2 {
        let sum: f64 = soft_caps.iter().map(|(_, h)| *h).sum();
        if sum > weekly_hours + 1e-9 {
            return false;
        }
    }
    for (tid, hours) in soft_caps {
        let used = per_goal.get(tid).copied().unwrap_or(0.0);
        if used > hours * 60.0 + 1e-9 {
            return false;
        }
    }
    // Weekly global with today only (dogfood same as JS when weekOther empty).
    if total > weekly_hours * 60.0 + 1e-9 {
        return false;
    }
    true
}

/// Deterministic propose-only plan under capacity.
pub fn propose_day_plan(input: &ProposeInput) -> ProposeOutput {
    if !input.weekly_capacity_hours.is_finite()
        || input.weekly_capacity_hours <= 0.0
        || !input.nights_per_week.is_finite()
        || input.nights_per_week <= 0.0
    {
        return ProposeOutput {
            ok: false,
            proposals: vec![],
            summary: "Set a positive weekly capacity and nights/week first.".into(),
            remaining_min: 0,
            error: Some("invalid_capacity".into()),
        };
    }

    let active: Vec<&TargetHint> = input.targets.iter().filter(|t| !t.id.is_empty()).collect();
    if active.is_empty() {
        return ProposeOutput {
            ok: false,
            proposals: vec![],
            summary: "Create at least one active target before AIly can propose a plan.".into(),
            remaining_min: daily_soft_cap_minutes(
                input.weekly_capacity_hours,
                input.nights_per_week,
            ) as u32,
            error: Some("no_targets".into()),
        };
    }

    let daily_cap =
        daily_soft_cap_minutes(input.weekly_capacity_hours, input.nights_per_week) as u32;
    let used: u32 = input
        .existing_today
        .iter()
        .filter(|c| c.status != "dropped")
        .map(|c| c.estimate_min)
        .sum();
    let mut remaining = daily_cap.saturating_sub(used);
    let max_items = input.max_items.clamp(1, 6);
    let intention = input.intention.trim();

    if remaining < FLOOR {
        return ProposeOutput {
            ok: true,
            proposals: vec![],
            summary: "Today is already full under your soft cap. Drop or replan before adding more."
                .into(),
            remaining_min: remaining,
            error: None,
        };
    }

    let mut ranked = active;
    ranked.sort_by(|a, b| {
        a.progress
            .partial_cmp(&b.progress)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| a.title.cmp(&b.title))
    });

    if !intention.is_empty() {
        let lower = intention.to_lowercase();
        if let Some(idx) = ranked.iter().position(|t| {
            t.title
                .to_lowercase()
                .split_whitespace()
                .any(|w| w.len() > 2 && lower.contains(w))
        }) {
            if idx > 0 {
                let item = ranked.remove(idx);
                ranked.insert(0, item);
            }
        }
    }

    let mut proposals: Vec<Proposal> = Vec::new();

    for t in ranked {
        if proposals.len() >= max_items || remaining < FLOOR {
            break;
        }
        let slots_left = max_items - proposals.len();
        let mut slice = clamp_estimate(remaining as f64 / slots_left as f64);
        if let Some(soft) = t.soft_hours.filter(|h| *h > 0.0) {
            let soft_day = clamp_estimate(soft * 60.0 / input.nights_per_week.max(1.0));
            slice = slice.min(soft_day);
        }
        slice = slice.min(90).min(remaining);
        slice = clamp_estimate(slice as f64);
        if slice < FLOOR {
            continue;
        }

        let text = if !intention.is_empty() && proposals.is_empty() {
            format!(
                "Protect: {}",
                intention.chars().take(80).collect::<String>()
            )
        } else {
            format!("Progress: {}", t.title.chars().take(60).collect::<String>())
        };
        let reason = if !intention.is_empty() && proposals.is_empty() {
            "Anchored to today’s intention".to_string()
        } else {
            format!(
                "Lowest progress first ({}% journey)",
                (t.progress * 100.0).round() as i32
            )
        };
        let must_keep = proposals.is_empty() && !intention.is_empty();

        let mut draft = Proposal {
            text,
            target_id: t.id.clone(),
            estimate_min: slice,
            must_keep,
            reason,
        };

        if !day_fits(
            input.weekly_capacity_hours,
            input.nights_per_week,
            &input.soft_caps,
            &input.existing_today,
            &proposals,
            &draft,
        ) {
            draft.estimate_min = FLOOR;
            if !day_fits(
                input.weekly_capacity_hours,
                input.nights_per_week,
                &input.soft_caps,
                &input.existing_today,
                &proposals,
                &draft,
            ) {
                continue;
            }
        }

        remaining = remaining.saturating_sub(draft.estimate_min);
        proposals.push(draft);
    }

    let total: u32 = proposals.iter().map(|p| p.estimate_min).sum();
    let summary = if proposals.is_empty() {
        "No safe proposals fit remaining capacity.".into()
    } else {
        format!(
            "Proposed {} block{} (~{}m) under your ~{}m day soft cap. Accept only what you mean.",
            proposals.len(),
            if proposals.len() == 1 { "" } else { "s" },
            total,
            daily_cap
        )
    };

    ProposeOutput {
        ok: true,
        proposals,
        summary,
        remaining_min: remaining,
        error: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn proposes_for_targets_under_cap() {
        let out = propose_day_plan(&ProposeInput {
            weekly_capacity_hours: 10.0,
            nights_per_week: 4.0,
            targets: vec![
                TargetHint {
                    id: "a".into(),
                    title: "Ship AIly".into(),
                    progress: 0.1,
                    soft_hours: Some(6.0),
                },
                TargetHint {
                    id: "b".into(),
                    title: "Health".into(),
                    progress: 0.8,
                    soft_hours: Some(4.0),
                },
            ],
            soft_caps: vec![("a".into(), 6.0), ("b".into(), 4.0)],
            existing_today: vec![],
            intention: "Ship AIly hard part".into(),
            max_items: 3,
        });
        assert!(out.ok);
        assert!(!out.proposals.is_empty());
        assert_eq!(out.proposals[0].target_id, "a");
        let total: u32 = out.proposals.iter().map(|p| p.estimate_min).sum();
        assert!(total <= 150);
    }

    #[test]
    fn empty_when_day_full() {
        let out = propose_day_plan(&ProposeInput {
            weekly_capacity_hours: 10.0,
            nights_per_week: 4.0,
            targets: vec![TargetHint {
                id: "a".into(),
                title: "Ship".into(),
                progress: 0.0,
                soft_hours: None,
            }],
            soft_caps: vec![],
            existing_today: vec![ExistingCommitment {
                id: "x".into(),
                target_id: "a".into(),
                estimate_min: 200,
                must_keep: false,
                status: "pending".into(),
            }],
            intention: String::new(),
            max_items: 3,
        });
        assert!(out.ok);
        assert!(out.proposals.is_empty());
    }

    #[test]
    fn requires_targets() {
        let out = propose_day_plan(&ProposeInput {
            weekly_capacity_hours: 10.0,
            nights_per_week: 4.0,
            targets: vec![],
            soft_caps: vec![],
            existing_today: vec![],
            intention: String::new(),
            max_items: 3,
        });
        assert!(!out.ok);
        assert_eq!(out.error.as_deref(), Some("no_targets"));
    }
}

File: apps/web/js/ally.js
/**
 * Local propose-only ally helpers (no cloud, no model download).
 * Deterministic suggestions the user can accept or ignore.
 */

import { checkPlanAccept, dailySoftCapMinutes } from "./capacity.js";
import { findSameDayDuplicate } from "./journey.js";
import { metricProgressRatio } from "./target.js";

const FLOOR = 15;

function clampEstimate(min) {
  if (!Number.isFinite(min) || min < FLOOR) return FLOOR;
  // Snap to 15m steps for dogfood friendliness.
  return Math.max(FLOOR, Math.round(min / 15) * 15);
}

/**
 * Existing Today ranking: must-keep first, then lower priority number,
 * then longer blocks. Same order the list and propose-adjacent pick use.
 */
export function rankCommitments(items) {
  return (Array.isArray(items) ? items : []).slice().sort((a, b) => {
    const mk = Number(!!b?.mustKeep) - Number(!!a?.mustKeep);
    if (mk) return mk;
    const pa = Number.isFinite(a?.priority) ? a.priority : 0;
    const pb = Number.isFinite(b?.priority) ? b.priority : 0;
    if (pa !== pb) return pa - pb;
    return (b?.estimateMin || 0) - (a?.estimateMin || 0);
  });
}

/**
 * Single next commitment the ally would start: first pending item
 * under the existing ranking.
 */
export function pickNextCommitment(items) {
  const pending = (Array.isArray(items) ? items : []).filter(
    (c) => c && c.status === "pending"
  );
  return rankCommitments(pending)[0] || null;
}

/**
 * Propose a capacity-honest set of commitments for today.
 * Never mutates state — pure.
 *
 * @param {{
 *   targets: Array,
 *   weeklyCapacityHours: number,
 *   nightsPerWeek: number,
 *   softCaps?: Array,
 *   existingToday?: Array,
 *   intention?: string,
 *   maxItems?: number,
 * }} input
 * @returns {{
 *   ok: boolean,
 *   proposals: Array<{ text: string, targetId: string, estimateMin: number, mustKeep: boolean, reason: string }>,
 *   summary: string,
 *   remainingMin: number,
 *   error?: string,
 * }}
 */
export function proposeDayPlan(input) {
  const targets = Array.isArray(input?.targets)
    ? input.targets.filter((t) => t && t.status === "active" && t.id)
    : [];
  const weekly = input?.weeklyCapacityHours;
  const nights = input?.nightsPerWeek;
  if (!Number.isFinite(weekly) || weekly <= 0 || !Number.isFinite(nights) || nights <= 0) {
    return {
      ok: false,
      proposals: [],
      summary: "Set a positive weekly capacity and nights/week first.",
      remainingMin: 0,
      error: "invalid_capacity",
    };
  }
  if (!targets.length) {
    return {
      ok: false,
      proposals: [],
      summary: "Create at least one active target before AIly can propose a plan.",
      remainingMin: dailySoftCapMinutes(weekly, nights),
      error: "no_targets",
    };
  }

  const dailyCap = dailySoftCapMinutes(weekly, nights);
  const existing = Array.isArray(input?.existingToday) ? input.existingToday : [];
  const used = existing.reduce(
    (a, c) => a + (c && c.status !== "dropped" && Number.isFinite(c.estimateMin) ? c.estimateMin : 0),
    0
  );
  let remaining = Math.max(0, dailyCap - used);
  const maxItems = Number.isFinite(input?.maxItems) ? Math.max(1, Math.min(6, input.maxItems)) : 3;
  const intention = typeof input?.intention === "string" ? input.intention.trim() : "";
  const softCaps = Array.isArray(input?.softCaps) ? input.softCaps : [];

  // Targets that already have pending work today — prefer other journeys first.
  const plannedTargetIds = new Set(
    existing
      .filter((c) => c && c.status === "pending" && c.targetId)
      .map((c) => c.targetId)
  );

  if (remaining < FLOOR) {
    return {
      ok: true,
      proposals: [],
      summary: "Today is already full under your soft cap. Drop or replan before adding more.",
      remainingMin: remaining,
    };
  }

  // Prefer targets furthest from their metric goal; stable by title for determinism.
  const ranked = targets
    .map((t) => ({
      target: t,
      progress: metricProgressRatio(t.metrics?.[0]),
      softHours:
        softCaps.find((s) => s.targetId === t.id)?.hours ??
        (Number.isFinite(t.softCapacityHours) ? t.softCapacityHours : null),
      alreadyPlanned: plannedTargetIds.has(t.id),
    }))
    .sort(
      (a, b) =>
        Number(a.alreadyPlanned) - Number(b.alreadyPlanned) ||
        a.progress - b.progress ||
        String(a.target.title).localeCompare(String(b.target.title))
    );

  const proposals = [];
  const intentionLower = intention.toLowerCase();

  // If intention mentions a target title, bias first slot toward that target.
  let order = ranked.slice();
  if (intention) {
    const hit = order.findIndex((r) =>
      String(r.target.title || "")
        .toLowerCase()
        .split(/\s+/)
        .some((word) => word.length > 2 && intentionLower.includes(word.toLowerCase()))
    );
    if (hit > 0) {
      const [item] = order.splice(hit, 1);
      order.unshift(item);
    }
  }

  const existingTexts = new Set(
    existing
      .filter((c) => c && c.status !== "dropped" && typeof c.text === "string")
      .map((c) => c.text.trim().toLowerCase().replace(/\s+/g, " "))
  );

  for (const row of order) {
    if (proposals.length >= maxItems || remaining < FLOOR) break;
    const t = row.target;
    // Skip targets that already have pending work unless intention named them first.
    if (row.alreadyPlanned && !(intention && proposals.length === 0 && order[0] === row)) {
      continue;
    }
    // Share remaining across remaining slots, but leave room for later targets.
    const slotsLeft = maxItems - proposals.length;
    let slice = clampEstimate(remaining / slotsLeft);

    // Respect soft weekly hours as a soft daily share when present.
    if (row.softHours != null && row.softHours > 0) {
      const softDay = clampEstimate((row.softHours * 60) / Math.max(1, nights));
      slice = Math.min(slice, softDay);
    }
    // Cap deep work blocks for dogfood readability.
    slice = Math.min(slice, 90);
    slice = clampEstimate(slice);
    if (slice > remaining) slice = clampEstimate(remaining);
    if (slice < FLOOR) continue;

    const metric = t.metrics?.[0];
    const metricHint = metric?.name ? ` on ${metric.name}` : "";
    const text = intention && proposals.length === 0
      ? `Protect: ${intention.slice(0, 80)}`
      : `Progress: ${String(t.title).slice(0, 60)}${metricHint}`;
    const normText = text.trim().toLowerCase().replace(/\s+/g, " ");
    if (existingTexts.has(normText) || proposals.some((p) => p.text.trim().toLowerCase().replace(/\s+/g, " ") === normText)) {
      continue;
    }

    const draft = {
      text,
      targetId: t.id,
      estimateMin: slice,
      mustKeep: proposals.length === 0 && !!intention,
      reason:
        proposals.length === 0 && intention
          ? "Anchored to today’s intention"
          : `Lowest progress first (${Math.round(row.progress * 100)}% journey)`,
    };

    const previewToday = [
      ...existing
        .filter((c) => c && c.status !== "dropped")
        .map((c) => ({
          id: c.id,
          targetId: c.targetId,
          estimateMin: c.estimateMin,
          mustKeep: !!c.mustKeep,
        })),
      ...proposals.map((p, i) => ({
        id: `prop-${i}`,
        targetId: p.targetId,
        estimateMin: p.estimateMin,
        mustKeep: p.mustKeep,
      })),
      {
        id: "prop-next",
        targetId: draft.targetId,
        estimateMin: draft.estimateMin,
        mustKeep: draft.mustKeep,
      },
    ];

    const check = checkPlanAccept({
      weeklyCapacityHours: weekly,
      nightsPerWeek: nights,
      softCaps,
      weekOther: [],
      today: previewToday,
    });
    if (!check.ok) {
      // Try smaller slice once.
      draft.estimateMin = FLOOR;
      previewToday[previewToday.length - 1].estimateMin = FLOOR;
      const retry = checkPlanAccept({
        weeklyCapacityHours: weekly,
        nightsPerWeek: nights,
        softCaps,
        weekOther: [],
        today: previewToday,
      });
      if (!retry.ok) continue;
    }

    proposals.push(draft);
    remaining -= draft.estimateMin;
  }

  // Always leave a small recovery buffer suggestion if room remains and we proposed deep work only.
  if (remaining >= 15 && proposals.length < maxItems && proposals.length > 0) {
    const buffer = clampEstimate(Math.min(15, remaining));
    const anchor = proposals[0];
    const draft = {
      text: "Buffer / break — protect recovery",
      targetId: anchor.targetId,
      estimateMin: buffer,
      mustKeep: false,
      reason: "Leave margin so the day stays honest",
    };
    const previewToday = [
      ...existing
        .filter((c) => c && c.status !== "dropped")
        .map((c) => ({
          id: c.id,
          targetId: c.targetId,
          estimateMin: c.estimateMin,
          mustKeep: !!c.mustKeep,
        })),
      ...proposals.map((p, i) => ({
        id: `prop-${i}`,
        targetId: p.targetId,
        estimateMin: p.estimateMin,
        mustKeep: p.mustKeep,
      })),
      {
        id: "prop-buffer",
        targetId: draft.targetId,
        estimateMin: draft.estimateMin,
        mustKeep: false,
      },
    ];
    const check = checkPlanAccept({
      weeklyCapacityHours: weekly,
      nightsPerWeek: nights,
      softCaps,
      weekOther: [],
      today: previewToday,
    });
    if (check.ok) {
      proposals.push(draft);
      remaining -= buffer;
    }
  }

  const summary = proposals.length
    ? `Proposed ${proposals.length} block${proposals.length === 1 ? "" : "s"} (~${proposals.reduce((a, p) => a + p.estimateMin, 0)}m) under your ~${Math.round(dailyCap)}m day soft cap. Accept only what you mean.`
    : "No safe proposals fit remaining capacity.";

  return {
    ok: true,
    proposals,
    summary,
    remainingMin: remaining,
  };
}

/**
 * Preview what accept-all would add or skip — never mutates.
 * Reasons: inactive | duplicate | capacity error from checkPlanAccept.
 *
 * @param {{
 *   proposals: Array,
 *   existingToday?: Array,
 *   allCommitments?: Array,
 *   targets: Array,
 *   weeklyCapacityHours: number,
 *   nightsPerWeek: number,
 *   softCaps?: Array,
 *   planDate: string,
 * }} input
 * @returns {{
 *   added: Array,
 *   skipped: Array<{ text?: string, targetId?: string, estimateMin?: number, reason: string }>,
 *   addedMin: number,
 *   remainingMin: number,
 *   dailyCapMin: number,
 * }}
 */
export function previewAcceptAll(input) {
  const proposals = Array.isArray(input?.proposals) ? input.proposals : [];
  const targets = Array.isArray(input?.targets) ? input.targets : [];
  const existingToday = Array.isArray(input?.existingToday) ? input.existingToday : [];
  const planDate = typeof input?.planDate === "string" ? input.planDate : "";
  const working = existingToday
    .filter((c) => c && c.status !== "dropped")
    .map((c) => ({
      id: c.id,
      targetId: c.targetId,
      estimateMin: c.estimateMin,
      mustKeep: !!c.mustKeep,
    }));
  const seen = Array.isArray(input?.allCommitments) ? input.allCommitments.slice() : existingToday.slice();
  const added = [];
  const skipped = [];

  for (const p of proposals) {
    if (!p) {
      skipped.push({ reason: "inactive" });
      continue;
    }
    const active = targets.some((t) => t && t.id === p.targetId && t.status === "active");
    if (!active) {
      skipped.push({ ...p, reason: "inactive" });
      continue;
    }
    const dup = findSameDayDuplicate(seen.concat(added), {
      planDate,
      text: p.text,
    });
    if (dup.duplicate) {
      skipped.push({ ...p, reason: "duplicate" });
      continue;
    }
    const draft = {
      id: `preview-${added.length}`,
      targetId: p.targetId,
      estimateMin: p.estimateMin,
      mustKeep: !!p.mustKeep,
      text: p.text,
      status: "pending",
      planDate,
    };
    const check = checkPlanAccept({
      weeklyCapacityHours: input?.weeklyCapacityHours,
      nightsPerWeek: input?.nightsPerWeek,
      softCaps: Array.isArray(input?.softCaps) ? input.softCaps : [],
      weekOther: [],
      today: [
        ...working,
        ...added.map((c) => ({
          id: c.id,
          targetId: c.targetId,
          estimateMin: c.estimateMin,
          mustKeep: !!c.mustKeep,
        })),
        {
          id: draft.id,
          targetId: draft.targetId,
          estimateMin: draft.estimateMin,
          mustKeep: draft.mustKeep,
        },
      ],
    });
    if (!check.ok) {
      skipped.push({ ...p, reason: check.error || "capacity" });
      continue;
    }
    added.push(draft);
  }

  const used = working.concat(added).reduce(
    (a, c) => a + (Number.isFinite(c.estimateMin) ? c.estimateMin : 0),
    0
  );
  const dailyCapMin = dailySoftCapMinutes(input?.weeklyCapacityHours, input?.nightsPerWeek);
  return {
    added,
    skipped,
    addedMin: added.reduce((a, c) => a + (Number.isFinite(c.estimateMin) ? c.estimateMin : 0), 0),
    remainingMin: Math.max(0, dailyCapMin - used),
    dailyCapMin,
  };
}

/**
 * Return-from-away check-in. A real question the UI can answer, plus
 * a single `text` line for logs / tests.
 * @param {{
 *   awayMin: number,
 *   intention?: string,
 *   focusActive?: boolean,
 *   openPending?: number,
 *   plannedMin?: number,
 * }} ctx
 * @returns {null | {
 *   question: string,
 *   text: string,
 *   awayMin: number,
 *   intention: string,
 *   focusActive: boolean,
 *   openPending: number,
 *   plannedMin: number,
 *   stillYes: string,
 *   stillNo: string,
 *   chooseNext: string,
 * }}
 */
export function returnNudge(ctx) {
  const away = Number.isFinite(ctx?.awayMin) ? ctx.awayMin : 0;
  if (away < 5) return null;
  const intention = typeof ctx?.intention === "string" ? ctx.intention.trim() : "";
  const open = Number.isFinite(ctx?.openPending) ? Math.max(0, Math.floor(ctx.openPending)) : 0;
  const planned = Number.isFinite(ctx?.plannedMin) ? Math.max(0, Math.round(ctx.plannedMin)) : 0;
  const awayRounded = Math.round(away);
  const planHint =
    open > 0
      ? ` You still have ${open} open item${open === 1 ? "" : "s"}${planned > 0 ? ` (~${planned}m planned)` : ""}.`
      : "";
  const question = intention
    ? `Still protecting ${intention.slice(0, 80)}?`
    : ctx?.focusActive
      ? "Is this still what you want to focus on?"
      : "What do you want the next stretch to be for?";
  let text;
  if (ctx?.focusActive) {
    text = `Welcome back (${awayRounded}m away). Focus is still on — is this still what you want?${planHint}`;
  } else if (intention) {
    text = `Welcome back (${awayRounded}m away). Your intention was “${intention.slice(0, 80)}”. Still true?${planHint}`;
  } else if (open > 0) {
    text = `Welcome back (${awayRounded}m away).${planHint} Pause — re-choose the next stretch intentionally.`;
  } else {
    text = `Welcome back (${awayRounded}m away). Pause — what do you want the next stretch to be for?`;
  }
  return {
    question,
    text,
    awayMin: awayRounded,
    intention,
    focusActive: !!ctx?.focusActive,
    openPending: open,
    plannedMin: planned,
    stillYes: intention ? "Yes — still this" : "Yes — keep going",
    stillNo: "No — change intention",
    chooseNext: "I'll choose next",
  };
}
